x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/perform_rule_upgrade/perform_rule_upgrade_route.schema.yaml

```diff
@@ -0,0 +1,507 @@
+openapi: 3.0.0
+info:
+  title: Perform Rule Upgrade API endpoint
+  version: '2023-10-31'
+paths:
+  /api/detection_engine/rules/prebuilt/_perform_upgrade:
+    post:
+      x-labels: [ess]
+      x-codegen-enabled: true
+      operationId: PerformRuleUpgrade
+      summary: Perform rule upgrade
+      description: Upgrade prebuilt detection rules.
+      tags:
+        - Rules API
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              oneOf:
+                - $ref: '#/components/schemas/UpgradeAllRulesRequest'
+                - $ref: '#/components/schemas/UpgradeSpecificRulesRequest'
+      responses:
+        200:
+          description: Indicates a successful call.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PerformRuleUpgradeResponseBody'
+
+components:
+  schemas:
+    RulePickVersionValues:
+      type: string
+      enum: [BASE, CURRENT, TARGET, MERGED]
+
+    FieldPickVersionValues:
+      type: string
+      enum: [BASE, CURRENT, TARGET, MERGED, RESOLVED]
+
+    RuleUpgradeSpecifier:
+      type: object
+      required:
+        - rule_id
+        - revision
```

Comment:
Why do we need to explicitly pass `revision` as a parameter? Is there a use case for it?

Reply:
Yes, it's a concurrency control check. We expect the `revision` to be passed here to match the latest existing in ES. If they do not match, that means there was a race condition, such as request slipping through between the /review and /perform calls, so that rule has already been updated or modified in some way. If that happens, the request for that rule should be rejected.